preview

schweitzpgi opened the pull request.

Commit:
Array expression lowering. Part 2.

Lowering of array assignments. This makes the old lowering of array expressions and assignments obsolete.

Expression lowering is bifurcated into the lowering of scalar expressions and array expressions (>= F90). This allows a natural nested scoping of sub-expression contexts as each sub-expression is lowered. The scalar expression code was mostly kept intact with some cleanup. The array expression lowering uses CPS (continuation passing style) to lower array expressions in parts: setup, loop nest and body, and cleanup.

While this is incomplete, it provides an interface and an implementation and an initial implementation for some target cases.

jeanPerier commented on flang/lib/Lower/ConvertExpr.cpp:
```diff
@@ -1697,133 +1625,486 @@ class ExprLowering {
+  // Lower the expression in a scalar context.
+  template <typename A>
+  ExtValue asScalar(const A &x) {
+    return ExprLowering{getLoc(), converter, symMap, stmtCtx}.genval(x);
```
I am not following by which magic this is computing the scalar outside of the loop (obviously, it is computed outside of the loops given your tests, so it works :) )

jeanPerier replied:
Ah, found it ! it is not being generated in a lamda :)

jeanPerier reviewed (approved):
Kudos, I really enjoyed reading at ConvertExpr.cpp! The continuation style passing really keeps the complexity down. I think it also looks very solid in terms of what kind of array expression it will be able to handle.
The only thing that I think might hit some problems is the cleanup. The overall statement approach is tempting but I fear it might bumps in subtle issues when the statement lowering generate control flow that may (especially when it some times does, but sometimes does not).
But there are probably ways to ensure that the clean-up at the statement level will work. I guess some experience with real code is needed here.

schweitzpgi commented on flang/lib/Lower/StatementContext.h:
```diff
@@ -0,0 +1,80 @@
+/// When lowering a statement, large temporaries may be allocated on the heap to
+/// buffer intermediate results. These temporaries must be deallocated at the
+/// end of the statement. These deallocations are threaded via a
+/// StatementContext back to the "end" of the statement.
```
Yep. We'll have to be very careful in order to get the cleanups in the correct positions. (Consider what I've done just a very rough draft in that regard. I expect it's buggy.)

Also I called it a "statement" context for convenience. It's really much more like a synchronization point in C. So we'll have to lower something like
```fortran
  IF (exp) 100, 200, 300
```
as
```
   eval exp
   cleanup any temps
   switch <, 100 ; =, 200 ; >,300
```
and not
```
   eval exp
   switch <, 100 ; =, 200 ; >,300
   cleanup any temps
```
obviously. :)

schweitzpgi commented on flang/lib/Lower/ConvertExpr.cpp:
```diff
@@ -1697,133 +1625,486 @@ class ExprLowering {
                             A, Fortran::evaluate::TypelessExpression>>>
-  void genpre(const A &) {}
+  CC genarr(const A &x) {
+    auto result = asScalar(x);
```
Hmm. I think this might be left over from some older work here.

vdonaldson committed:
Rewrite "if (cond) goto N" statements to eliminate the goto (#560)

* Rewrite "if (cond) goto N" statements to eliminate the goto

Rewrite:

    if (cond) goto L
    <statement1>
    <statement2>
  L <statement3>

or:

    if (cond) then
      goto L
    endif
    <statement1>
    <statement2>
  L <statement3>

as:

    if (.not. cond) then
      <statement1>
      <statement2>
    endif
  L <statement3>

This transformation eliminates a trivial basic block containing just
a goto, and allows implementation of the IF as a structured fir.if
(assuming that the statements that are moved into the IF construct
are themselves structured), enabling additional code optimization.

There are three places where this transformation could be done.
All three have problems.

1. In the front end, similar to canonicalization of DO statements.

Primary problem: This would transform some conformant programs into
nonconformant programs that branch into the body of the resulting
IF construct, complicating parse-unparse-reparse code sequences.
(Nonconformance does not itself compromise correct code generation.)

2. When building the PFT.

Primary problem: This would require changing the parse tree, which
is otherwise read-only.

3. When generating FIR from the PFT.

Primary problem:  This transformation invalidates some of the analysis
done when building the PFT.  That analysis would need to be redone
to avoid generation of incorrect code.

There are possible ways to mitigate each of these problems, such as
restricting the number of cases that are transformed, implementing
additional analyses, or copying code.  However, there are additional
problems in some cases, and more importantly, there is an alternative.

The alternative is to use a hybrid approach that splits the
implementation into two or three parts to avoid these problems.
The primary split is to make structural control flow changes when
building the PFT, but defer expression level changes to FIR generation.
The structural changes can be done without modifying the parse tree,
which avoids the option #2 problem.  When done at the right time,
these changes are sufficient to allow correct analysis of the
transformed IR, which avoids the option #3 problem.

Since at least some IF statements are rewritten as IF constructs,
there is a question as to whether it might be useful to rewrite all
IF statements.  This does not by itself change generated code.
However, it has several secondary benefits, including simple enabling
of the "if (cond) then; goto L; endif" case, so this PR does that
conversion for all cases.

That splits the implementation into three parts.  All IF statements
are converted to IF constructs when building the PFT.  After the
basic PFT is built, but before branches are analyzed, IF constructs
containing just a GOTO are visited, and viable candidates are
transformed to eliminate the GOTO.  This is effectively a new pass.
IfStmt's and IfThenStmt's that are transformed are tagged with a
"negateCondition" flag.  After branch analysis finalizes the PFT,
FIR generation looks at this flag to determine when to negate a
branch condition.

A key part of this implementation is the selection of candidates
to transform.  The transformation is only valid for forward branch
targets at the same construct nesting level as the IfConstruct.
The result must not violate construct nesting requirements or contain
an EntryStmt.  The result is allowed to violate the F18 Clause 11.1.2.1
prohibition on transfer of control into the interior of a construct
block, as that does not compromise correct code generation.  When two
transformation candidates overlap, at least one must be disallowed.
In such cases, the current heuristic favors simple code generation,
which happens to favor later candidates over earlier candidates.
That choice is probably not significant, but could be changed.

Commit:
review comments
conflicts
